<?xml version="1.0"?>
<launch>
    <!-- rviz -->
    <node name="rvizualizer" pkg="rviz" type="rviz" args="-d $(find dq_ipp)/../rviz.rviz"/>

    <!-- joystick -->
    <node name="ctrl_joy" pkg="dq_ipp" type="mavros_joy_controller.py" output="screen"/>
    
    <node name="joy_ros" pkg="joy" type="joy_node">
        <param name="/dev" value="/dev/input/js0"/>
    </node>
    
    <!-- Gazebo configs -->
    <arg name="world" default="$(find dq_ipp)/../gazebo_env/emart.world"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.01"/>
    <arg name="R" default="0"/>
    <arg name="P" default="0"/>
    <arg name="Y" default="3.141592"/>
    <arg name="gui" default="false"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="false"/>
    <arg name="paused" default="false"/>
    <arg name="respawn_gazebo" default="false"/>

    <!-- vehicle configs -->
    <arg name="est" default="ekf2"/>
    <arg name="vehicle" default="dji_matrice_100"/>
    <arg name="sensor" default="sensored_m100"/>
    <arg name="sdf" default="$(find dq_ipp)/../gazebo_env/$(arg sensor)/$(arg sensor).sdf"/>

    <!-- MAVROS configs -->
    <arg name="fcu_url" default="udp://:14540@localhost:14557"/>
    <arg name="respawn_mavros" default="false"/>
    <!-- PX4 configs -->
    <arg name="interactive" default="true"/>
    <!-- Voxblox configs -->
    <arg name="voxel_size" default="0.2"/>
    <arg name="voxels_per_side" default="16"/>
    <arg name="ray_length" default="5.0"/>
    <arg name="method" default="merged"/>
    <arg name="verbose_voxblox" default="false"/>


    <!-- tf broadcaster is needed -->
    <node name="tf_broader" pkg="dq_ipp" type="tf_broadcaster.py" output="screen"/>

    <!-- rgbd node -->
    <node name="rgbd_node" pkg="dq_ipp" type="rgbd_node" output="screen">
        <rosparam param="/cam_intrinsic">[640, 480, 319.998, 319.998, 320.5, 240.5]</rosparam> <!-- width, height, f_x, f_y, c_x, c_y -->
        <rosparam param="/cam_extrinsic">[0.1, 0.0, 0.25]</rosparam> <!-- body to cam: translation x, y, z, downward pitch angle (rad) -->
    </node>


    <!-- PX4 SITL and Gazebo -->
    <include file="$(find px4)/launch/posix_sitl.launch">
        <arg name="x" value="$(arg x)"/>
        <arg name="y" value="$(arg y)"/>
        <arg name="z" value="$(arg z)"/>
        <arg name="R" value="$(arg R)"/>
        <arg name="P" value="$(arg P)"/>
        <arg name="Y" value="$(arg Y)"/>
        <arg name="world" value="$(arg world)"/>
        <arg name="vehicle" value="$(arg vehicle)"/>
        <arg name="sdf" value="$(arg sdf)"/>
        <arg name="gui" value="$(arg gui)"/>
        <arg name="interactive" value="$(arg interactive)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="respawn_gazebo" value="$(arg respawn_gazebo)"/>
    </include>
    <!-- MAVROS -->
    <include file="$(find mavros)/launch/px4.launch">
        <!-- GCS link is provided by SITL -->
        <arg name="gcs_url" value=""/>
        <arg name="fcu_url" value="$(arg fcu_url)"/>
        <arg name="respawn_mavros" value="$(arg respawn_mavros)"/>
    </include>

    <!-- Voxblox -->
    <node name="voxblox_node" pkg="voxblox_ros" type="esdf_server" output="screen" args="-alsologtostderr">
        <!-- <remap from="pointcloud" to="/dji_matrice_100/camera/depth/points"/> -->
        <remap from="pointcloud" to="/aligned_points"/>
        <param name="use_tf_transforms" value="true"/>
        <param name="publish_tsdf_map" value="true" />
        <param name="publish_esdf_map" value="true" />
        <param name="publish_traversable" value="false" />
        <param name="publish_pointclouds" value="false" />
        <param name="tsdf_voxel_size" value="$(arg voxel_size)" />
        <param name="tsdf_voxels_per_side" value="$(arg voxels_per_side)" />
        <param name="color_mode" value="color" />
        <param name="update_mesh_every_n_sec" value="1.0" />
        <param name="method" value="$(arg method)" />
        <param name="max_ray_length_m" value="$(arg ray_length)" />
        <param name="allow_clear" value="true" />
        <param name="verbose" value="$(arg verbose_voxblox)" />
    </node>

    <!-- Planner -->
    <node name="planner_node" pkg="dq_ipp" type="planner_node" output="screen">
        <remap from="planner_node/esdf_map_in" to="voxblox_node/esdf_map_out" />
        <remap from="planner_node/tsdf_map_in" to="voxblox_node/tsdf_map_out" />
        <param name="use_tf_transforms" value="false"/>
        <param name="tsdf_voxel_size" value="$(arg voxel_size)" />
        <param name="tsdf_voxels_per_side" value="$(arg voxels_per_side)" />
        <param name="update_mesh_every_n_sec" value="0.0" /> 
        <param name="publish_traversable" value="true" />
        <param name="publish_pointclouds" value="true" />

        <!-- planner.cpp -->
        <param name="p_verbose" value="true" />
        <!-- voxblox.cpp -->
        <param name="p_collision_radius" value="0.1" />
        <!-- ray_caster.cpp -->
        <param name="p_downsampling_factor" value="2.0" />
        <param name="p_ray_length" value="5.0" />
        <param name="p_focal_length" value="320.0" />
        <param name="p_resolution_x" value="640.0" />
        <param name="p_resolution_y" value="480.0" />
        <!-- frontier.cpp -->
        <param name="p_verbose_ft" value="false" />
        <param name="p_down_sample" value="1" />
        <param name="p_spatial_cluster_min" value="50" />
        <param name="p_surface_cluster_min" value="5" />
        <param name="p_cluster_size_xy" value="2.0" />
        <param name="p_cluster_size_yz" value="2.0" />
        <param name="p_vp_min_visible_num" value="15.0" />
        <param name="p_vp_rmax" value="4.5" />
        <param name="p_vp_rmin" value="3.5" />
        <param name="p_vp_rnum" value="3" />
        <param name="p_vp_dphi" value="$(eval 15 * 3.1415926 / 180.0)" />
        <param name="p_vp_min_dist" value="0.4" />
        <param name="p_vp_clearance" value="0.41" />

        <!-- rrt star -->
        <param name="max_rrt_iteration" value="20000" />
        <param name="rrt_collision_r" value="1.0" />
        <param name="extension_range" value="1.0" />
        <rosparam param="map_min"> [-25.0, -15.0, 0.5] </rosparam>
        <rosparam param="map_max"> [10.0, 15.0, 20.0] </rosparam>
        <rosparam param="map_min2"> [-40.0, -40.0, 0.5] </rosparam>
        <rosparam param="map_max2"> [25.0, 40.0, 50.0] </rosparam>

        <param name="p_gain" value="0.8" />
        <param name="p_gain_yaw" value="2.0" />
        <param name="linear_vel_bound" value="1.0" />
        <param name="yaw_bound" value="1.57" /> <!-- 2.355: 1.5pi -> non perfect mesh -->
        <param name="init_rot_sec" value="6.0" />
        <param name="reached_pos_th" value="0.7" />
        

        <remap from="odometry" to="/mavros/local_position/pose" />
        <!-- <remap from="command/trajectory" to="command/trajectory" /> -->

    </node>

    <!-- Evaluate -->
    <node name="evaluate" pkg="dq_ipp" type="evaluate.py" output="screen">
        <rosparam param="map_min"> [-25.0, -33.0, 0.5] </rosparam>
        <rosparam param="map_max"> [10.0, 33.0, 20.0] </rosparam>
        <param name="voxel_size" value="0.2"/>
    </node>

    <!-- Visualization -->
    <group ns="dji_matrice_100">
        <node name="pose_to_path" pkg="tf_to_trajectory" type="posestamped_to_path.py" output="screen">
            <param name="frame_id" value="map"/>
            <param name="append_rate" value="3"/>
            <param name="topic_name" value="/mavros/local_position/pose"/>
        </node>
    </group>
</launch>